DataFormats/Detectors/TRD/include/DataFormatsTRD/TrapConfig3.h

```diff
@@ -0,0 +1,619 @@
+  uint32_t getRegisterValue(uint32_t regidx, int mcmidx);
+  bool setRegisterValue(uint32_t data, uint32_t regidx, int mcmidx);
+  uint32_t getRegisterValueByIdx(uint32_t regidx, int mcmidx);
+  bool setRegisterValueByIdx(uint32_t data, uint32_t regidx, int mcmidx);
+  uint32_t getRegisterValueByAddr(uint32_t addr, int mcmidx);
+  bool setRegisterValueByAddr(uint32_t data, uint32_t addr, int mcmidx);
+  uint32_t getRegisterValueByName(const std::string& name, int mcmidx);
+  bool setRegisterValueByName(uint32_t data, const std::string& regname, int mcmidx);
+
+  // get a config register value by index, addr, and name, via sector/stack/layer/rob/mcm
+  // no setters for sector/stack/layer/rob/mcm as writing, we only write from retrieved config events.
+  uint32_t getRegisterValueByIdx(uint32_t regix, int sector, int stack, int layer, int rob, int mcm);
+  uint32_t getRegisterValueByAddr(uint32_t regaddr, int sector, int stack, int layer, int rob, int mcm);
+  uint32_t getRegisterValueByIdx(uint32_t regix, int detector, int rob, int mcm);
+  uint32_t getRegisterValueByAddr(uint32_t regaddr, int detector, int rob, int mcm);
+
+  // get a registers name by addres and index;
+  std::string getRegNameByAddr(uint16_t addr);
+  std::string getRegNameByIdx(unsigned int regidx);
+
+  // get a registers index (enum value) by address or name
+  int32_t getRegIndexByAddr(unsigned int addr);
+  int32_t getRegIndexByName(const std::string& name);
+
+  // get a registers address by index or name
+  int32_t getRegAddrByIdx(unsigned int regidx);
+  int32_t getRegAddrByName(const std::string& name);
+
+  // retrieve the trap register info for a particular register, no need to index into det/rob/mcm as these are constant over a config
+  o2::trd::TrapRegInfo* getTrapRegInfoByAddr(uint32_t addr);
+  o2::trd::TrapRegInfo* getTrapRegInfoByIdx(uint32_t idx);
+
+  // get the indenfiying characteristics of a register given its address.
+  void getRegisterByAddr(uint32_t registeraddr, std::string& regname, int32_t& newregidx, int32_t& numberbits);
+
+  // return all the registers for a particular mcm
+  void getAllRegisters(int mcmidx, std::array<uint32_t, kLastReg>& registers);
+
+  // return all the mcm values for a particular register index
+  void getAllMCMByIndex(int regindex, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& registers);
+
+  // return all the mcm values for a particular register name
+  void getAllMCMByName(std::string registername, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the mcm for a particular register address
+  void getAllMCMByAddress(int registeraddress, std::array<uint32_t, o2::trd::constants::MAXMCMCOUNT>& mcms);
+
+  // return all the config data in an unpacked array.
+  void getAll(std::array<uint32_t, kLastReg * o2::trd::constants::MAXMCMCOUNT>& configdata);
+
+  // return the full unpacked config event.
+  bool printRegister(int addr, int det, int rob, int mcm);
```

Comment:
I think you would want to call this with a register index, i.e.:
```
tc.printRegister(kADCMSK, getMCMId(det,rob,mcm));
```
In that case, a single method would be sufficient.

Reply:
sure